pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,408 @@
+		for _, nodeID := range block.NUMAAffinity {
+			nodes = append(nodes, strconv.Itoa(nodeID))
+		}
+	}
+
+	if len(nodes) < 1 {
+		klog.V(5).Infof("[memorymanager] update container resources is skipped due to memory blocks are empty")
+		return nil
+	}
+
+	affinity := strings.Join(nodes, ",")
+	klog.Infof("[memorymanager] Set container %q cpuset.mems to %q", containerID, affinity)
+	err := m.containerRuntime.UpdateContainerResources(containerID, &runtimeapi.LinuxContainerResources{CpusetMems: affinity})
+	if err != nil {
+		klog.Errorf("[memorymanager] AddContainer error: error updating cpuset.mems for container (pod: %s, container: %s, container id: %s, err: %v)", pod.Name, container.Name, containerID, err)
+
+		m.Lock()
+		err = m.policyRemoveContainerByRef(string(pod.UID), container.Name)
+		if err != nil {
+			klog.Errorf("[memorymanager] AddContainer rollback state error: %v", err)
+		}
+		m.Unlock()
+	}
+	return err
+}
+
+// Allocate is called to pre-allocate memory resources during Pod admission.
+func (m *manager) Allocate(pod *v1.Pod, container *v1.Container) error {
+	// Garbage collect any stranded resources before allocation
+	m.removeStaleState()
+
+	m.Lock()
+	defer m.Unlock()
+
+	// Call down into the policy to assign this container memory if required.
+	if err := m.policy.Allocate(m.state, pod, container); err != nil {
+		klog.Errorf("[memorymanager] Allocate error: %v", err)
+		return err
+	}
+	return nil
+}
+
+// RemoveContainer removes the container from the state
+func (m *manager) RemoveContainer(containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// if error appears it means container entry already does not exist under the container map
+	podUID, containerName, err := m.containerMap.GetContainerRef(containerID)
+	if err != nil {
+		klog.Warningf("[memorymanager] Failed to get container %s from container map error: %v", containerID, err)
+		return nil
+	}
+
+	err = m.policyRemoveContainerByRef(podUID, containerName)
+	if err != nil {
+		klog.Errorf("[memorymanager] RemoveContainer error: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// State returns the state of the manager
+func (m *manager) State() state.Reader {
+	return m.state
+}
+
+// GetPodTopologyHints returns the topology hints for the topology manager
+func (m *manager) GetPodTopologyHints(pod *v1.Pod) map[string][]topologymanager.TopologyHint {
+	// Garbage collect any stranded resources before providing TopologyHints
+	m.removeStaleState()
+	// Delegate to active policy
+	return m.policy.GetPodTopologyHints(m.state, pod)
+}
+
+// GetTopologyHints returns the topology hints for the topology manager
+func (m *manager) GetTopologyHints(pod *v1.Pod, container *v1.Container) map[string][]topologymanager.TopologyHint {
+	// Garbage collect any stranded resources before providing TopologyHints
+	m.removeStaleState()
+	// Delegate to active policy
+	return m.policy.GetTopologyHints(m.state, pod, container)
+}
+
+func (m *manager) removeStaleState() {
```

Comment:
just noting how a lot of the logic maps to what we have in cpu manager.  we may want to find a factoring in future that allows us to move the sources ready stuff up a level to meta manager of managers.  for now, this is fine, but the less duplication we have in long run the better.

Reply:
You totally right, I will add it to my TODO list

Reply:
+1

Though I don't want to simply merge the two into a single Manager. If we can find a good way to abstract out the common bits, but still have distinct managers for each, that would be ideal.